### PR TITLE
Adding manage_dlm privilege

### PR DESCRIPTION
**Pull request by masseyke**
This adds a new index privilege called `manage_dlm`. The `manage_dlm` has permission to perform all DLM actions on an index, including put and delete. It also adds the ability to call DLM get and explain to the `view_index_metadata` existing index privilege.

### CODE REVIEW

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java
Comment (andreidan) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java:71]: Do we ... expect an unknown index privilege message error after introducing a new privilege?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
Comment (andreidan) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java:129]: This shows we have a bit of an inconsistency in our action names. I'll bring the `explain` API into the `data_lifecycle` namespace (this comment is just a remark in general :) nothing about this PR)

**Review by andreidan — COMMENTED** [review]
Thanks for working on this Keith, it's looking great so far.

Just a few questions:

Do we need to update `ClusterPrivilegeResolver` too ?

Do we need documentation update? i.e. `get-builtin-privileges.asciidoc` 

I think the docs for all DLM related APIs need to have this security instruction:
```
* If the {es} {security-features} are enabled, you must have the `manage_dlm` or
`read_ilm` or both cluster privileges to use this API. For more information, see
<<security-privileges>>.
```

I'd say we should give `managed_dlm` to the relevant builtin roles defined in `ReservedRolesStore` ? (at least `kibanaSystemRoleDescriptor` ?)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java:129]: Opened https://github.com/elastic/elasticsearch/pull/95555 and actually brought all actions under the `dlm` namespace. 
If `data_lifecycle` is preferable we can go that way ?

**Comment by masseyke** [comment]
> Do we need to update `ClusterPrivilegeResolver` too ?

I don't believe so. We don't have any cluster-level actions we're protecting do we? ILM had some but DLM does not. I had originally added a cluster-level privilege and then realized I had no need for it so I removed it.

**Comment by masseyke** [comment]
> Do we need documentation update? i.e. `get-builtin-privileges.asciidoc`

For some reason I was thinking we had discussed this as a follow-up PR, but I can't find that conversation so maybe I'm thinking of something else! I'll add some docs on this.

**Comment by masseyke** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> I'd say we should give `managed_dlm` to the relevant builtin roles defined in `ReservedRolesStore` ? (at least `kibanaSystemRoleDescriptor` ?)

Those are cluster privileges, and this is an index-level privilege. I actually had this one in there previously, but after I removed it as a cluster-level privilege it caused tests to start breaking if it remained in ReservedRolesStore.

**Comment by n1v0lg** [comment]
@masseyke I'm picking up DLM security model related work (starting next week I'll be on the project full time), so I'll review this from the es-security side. I'll get a review in by EOD tomorrow as I'm still catching up on context.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java
Comment (n1v0lg) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java:71]: The test here pertains to backwards compatibility, specifically for API keys. For API keys, we serialize the role descriptors (as XContent) associated with it (which includes index privileges), and store those as a doc in the security index. The reason the test is failing here is because the test runs on an old version that doesn't have `manage_dlm` defined as a privilege yet. Further up in the test, we are creating a random role descriptor with a set of random index privileges, that sometimes include `manage_dlm` leading to the unknown index privilege failure here.

Just wanted to leave this as context, to save a bit of digging.

In terms of addressing this, for the context of this PR, I think it's all good to keep this as is. I might follow this up next week to make things a little more robust in a mixed-cluster scenario but it's too much of a side quest for the current PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java:129]: Thanks Keith, I merged https://github.com/elastic/elasticsearch/pull/95555 so let's update this to just use the `dlm` namespace

**Review by andreidan — APPROVED** [review]
Thanks for working on this Keith 🚀 

This LGTM pending @n1v0lg's review.

**Comment by andreidan** [comment]
@masseyke @n1v0lg I _think_ we'd have to give fleet `manage_dlm` to be able to install packages with the DLM configuration - should they choose to do so (these would be changes to the managed index templates) but the index template security model is pending so we can do that later I believe. Just wanted to make a note of this.
 
https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java#L802

**Comment by n1v0lg** [comment]
Regarding the role descriptors in `ReservedRoleStore`, @masseyke it's true that there are cluster privileges, but there are also index privileges, e.g.,: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java#L836

Agreed with @andreidan's [comment](https://github.com/elastic/elasticsearch/pull/95512#issuecomment-1525510202) though, I think it only makes sense to update those once we've finalized the design around the security model.

I'm wrapping up the review, just want to sanity check a few things with @jakelandis around the overall work for the security model changes before giving this a ✅.

**Comment by n1v0lg** [comment]
@masseyke @andreidan having caught up on the initial plans for the security model, I think we might need to hold off on this change for a bit, as it depends on the exact approach we take for DLM security overall. 

The current thinking for the security model is, at a high level, to have a cluster privilege check on the `GET | PUT | DELETE /_data_stream/{name}/_lifecycle` APIs to ensure that a user is generally permitted to interact with DLM (this would be granted with a cluster privilege `manage_dlm`) and additionally check that the user has the requisite index privileges to perform the index-level actions (e.g., delete) that might be triggered by DLM, over the specific indices or data streams. 

As such, we (most likely) want to change e.g., `PutDataLifecycleAction` from an index action to a cluster action, add a new cluster privilege (`manage_dlm`), and then rely on additional code for checking concrete index-level access (how this will work is part of designing the DLM security model, which I'm picking up imminently). 

This is just the current thinking, and could change as I work out the design for DLM security. Once I've made progress, I will share here ASAP. I'm also more than happy to take over the PR and adjust as needed. @masseyke thank you for working on this! Apologies for not getting this review out sooner. I only just hopped on this project, so there is still quite a bit of context I have to gather.

**Comment by n1v0lg** [comment]
> I believe that this is derived from the action names. Since all of the DLM ones begin with indices:admin (and are actions on indices or data streams) we use index-level privileges. 

This part is a little trickier: for all actions we first check if they are cluster actions [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L436) which under the hood checks if the action name is part of the [`ALL_CLUSTER_PATTERN`](https://github.com/elastic/elasticsearch/blob/e752135f4446056a67ca6fcd7e246efbfa87efce/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ClusterPrivilegeResolver.java#L104). This means that we can have an action of the form `indices:...` that is still treated and authorized as a cluster action. We will likely want to do this for the data lifecycle actions so they get authorized against cluster privileges (pending more work on the security model design to say this for sure).

**Comment by masseyke** [comment]
One note before putting this PR aside -- the latest test failure appears to have turned up a real bug (https://gradle-enterprise.elastic.co/s/6kb67wslyurna/tests/:x-pack:qa:rolling-upgrade:v7.17.10%23oldClusterTest/org.elasticsearch.upgrades.ApiKeyBackwardsCompatibilityIT/testCreatingAndUpdatingApiKeys?expanded-stacktrace=WyIwIl0&top-execution=1). I've been talking with @slobodanadamovic about it. If you are running in a mixed mode cluster, and if your newer nodes have a new index privilege, and if you create an API key with this new privilege and use it to make a request that goes to an old cluster, then your request fails. I think Slobodan is going to write up a ticket and look into this. I just thought I'd mention it here so that no one spends too much time on that test failure -- it's not a bug in this PR.

**Comment by n1v0lg** [comment]
@masseyke thanks for raising this! 

I alluded (quite vaguely) to this issue in my comment here:

> I might follow this up next week to make things more robust in a mixed-cluster scenario but it's too much of a side quest for the current PR.

I was only thinking about the new privilege we're adding here, but @slobodanadamovic is correct in that it doesn't just apply to the new privilege, but more generally to other privileges we've added in the past, so it makes sense to track it as a bug beyond the scope of this current PR. Good catch 👍

**Comment by slobodanadamovic** [comment]
@n1v0lg @masseyke 

I took some time to look into the issue with BWC and it turns out that it's not only affecting API Keys but also Users.

Creating a role with the newly introduced index privilege will cause problems in a mixed cluster. The same problem can happen when adding a new cluster privileges. I think we need to come up with a generic way of handling BWC for index and cluster privileges.

I've reported a bug here: https://github.com/elastic/elasticsearch/issues/95771

**Comment by n1v0lg** [comment]
Reviving this, since it's the final approach we settled on after iterating on design. 

I'll leave a review with a few small comments and nits. @masseyke I'm more than happy to push those changes to this PR, or let you address them. Let me know which you prefer! Thanks for working on this.

File: modules/dlm/qa/with-security/build.gradle
Comment (n1v0lg) [inline modules/dlm/qa/with-security/build.gradle:16]: To account for the privilege sitting behind a feature flag, I think we need:

```suggestion
  user username: "test_non_privileged", password: "x-pack-test-password", role: "not_privileged"
  requiresFeature 'es.dlm_feature_flag_enabled', Version.fromString("8.9.0")
```

File: modules/dlm/qa/with-security/src/javaRestTest/java/org/elasticsearch/dlm/PermissionsIT.java
Comment (n1v0lg) [inline modules/dlm/qa/with-security/src/javaRestTest/java/org/elasticsearch/dlm/PermissionsIT.java:82]: Would be great to cover a 403 scenario for a data stream not covered by index pattern in role, e.g.,:

```suggestion
        String otherDataStreamName = "other-dlm-test";
createDataStreamAsAdmin(otherDataStreamName);
Response getOtherDataStreamResponse = adminClient().performRequest(new Request("GET", "/_data_stream/" + otherDataStreamName));
final List<Map<String, Object>> otherNodes = ObjectPath.createFromResponse(getOtherDataStreamResponse).evaluate("data_streams");
String otherIndex = (String) ((List<Map<String, Object>>) otherNodes.get(0).get("indices")).get(0).get("index_name");
Request putOtherLifecycleRequest = new Request("PUT", "_data_stream/" + otherDataStreamName + "/_lifecycle");
putOtherLifecycleRequest.setJsonEntity("{}");
makeRequest(client(), new Request("POST", "/" + otherIndex + "/_lifecycle/explain"), false);
makeRequest(client(), new Request("GET", "_data_stream/" + otherDataStreamName + "/_lifecycle"), false);
makeRequest(client(), new Request("DELETE", "_data_stream/" + otherDataStreamName + "/_lifecycle"), false);
makeRequest(client(), putOtherLifecycleRequest, false);
        }
```

**Review by n1v0lg — COMMENTED** [review]
Thanks for working on this! Just a few smaller comments. I can address these myself too, if you're swamped. Let me know what you prefer.

**Comment by masseyke** [comment]
> Reviving this, since it's the final approach we settled on after iterating on design.
> 
> I'll leave a review with a few small comments and nits. @masseyke I'm more than happy to push those changes to this PR, or let you address them. Let me know which you prefer! Thanks for working on this.

I've got time to address them today. Thanks.

File: modules/dlm/qa/with-security/src/javaRestTest/java/org/elasticsearch/dlm/PermissionsIT.java
Comment (masseyke) [reply modules/dlm/qa/with-security/src/javaRestTest/java/org/elasticsearch/dlm/PermissionsIT.java:67]: We've only granted permission on `dlm-*` though right? So we'd expect it to potentially fail for others. Or are you suggesting I change the test to make sure that it _does_ fail for others?

Comment (n1v0lg) [reply modules/dlm/qa/with-security/src/javaRestTest/java/org/elasticsearch/dlm/PermissionsIT.java:67]: No sorry, I'm suggesting it will work for these patterns as well: the way that index name resolution works is that `_all` and "*" will resolve into _only_ the names the users has access for (so essentially `dataStreamName`)

Comment (masseyke) [reply modules/dlm/qa/with-security/src/javaRestTest/java/org/elasticsearch/dlm/PermissionsIT.java]: Oh I didn't realize that. Thanks.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
Comment (masseyke) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java:211]: Weird -- I could have sworn I had done this, but clearly I hadn't!

File: modules/dlm/qa/with-security/src/javaRestTest/java/org/elasticsearch/dlm/PermissionsIT.java
Comment (n1v0lg) [reply modules/dlm/qa/with-security/src/javaRestTest/java/org/elasticsearch/dlm/PermissionsIT.java]: Sure thing! Only learned this recently also.

File: modules/dlm/qa/with-security/roles.yml
Comment (masseyke) [reply modules/dlm/qa/with-security/roles.yml:9]: I had it in there just so that it was very clear that we're just removing one privilege for the 'not_privileged` role. But I don't mind removing it.

File: x-pack/docs/en/rest-api/security/get-builtin-privileges.asciidoc
Comment (n1v0lg) [reply x-pack/docs/en/rest-api/security/get-builtin-privileges.asciidoc]: Oh ya good q! I took another look: I think we should remove it from the list above **and** also remove the `manage_dlm` docs portion from `privileges.asciidoc` (I'll add a separate suggestion in a comment to remove it).

The reason is that docs don't work well with feature flags when it comes to pages that are themselves not feature flagged but have sections that need to be feature flagged. AFAIK there is no way to conditionally skip skip a section of a page, so I think it's best to omit the section from docs in such cases. Otherwise, we will end up with docs for a privilege that isn't actually available to end-users yet.

I'll create a separate Jira issue to complete docs for DLM privileges, so we don't lose track of the docs gap as we get closer to removing the feature flag.

Note that the DLM-specific pages are fine, and we can reference `manage_dlm` there already. Those are marked experimental and will therefore not show up in our "real" docs AFAIK.

File: docs/changelog/95512.yaml
Comment (n1v0lg) [inline docs/changelog/95512.yaml:4]: I'm wondering if we really want a changelog here: the privilege is behind a feature flag, and so is the DLM feature itself. I don't know if we've been including changelogs for DLM so far but it might look weird to users to see changelogs for something that isn't available outside a feature flag yet. We could instead mark this `non-issue` and skip the changelog.

File: x-pack/docs/en/security/authorization/privileges.asciidoc
Comment (n1v0lg) [inline x-pack/docs/en/security/authorization/privileges.asciidoc:306]: I'd remove this whole section for now. See my other [comment](https://github.com/elastic/elasticsearch/pull/95512#discussion_r1206397157).

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java
Comment (n1v0lg) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java:90]:
```suggestion
                           
```

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java
Comment (n1v0lg) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java:117]:
```suggestion
                                
```

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java
Comment (n1v0lg) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java:127]:
```suggestion
                            
```

**Review by n1v0lg — APPROVED** [review]
LGTM! Few smaller things, but no need for another review from my end.

File: docs/changelog/95512.yaml
Comment (n1v0lg) [reply docs/changelog/95512.yaml:4]: No strong opinion on my end!

File: x-pack/docs/en/rest-api/security/get-builtin-privileges.asciidoc
Comment (masseyke) [reply x-pack/docs/en/rest-api/security/get-builtin-privileges.asciidoc]: I created #96386 so that we don't forget to change this in the future.

Comment (andreidan) [reply x-pack/docs/en/rest-api/security/get-builtin-privileges.asciidoc]: Hm, I'm curious if the tests will pass if we don't mention it - note that the feature flag is always enabled for SNAPSHOT builds

Comment (n1v0lg) [reply x-pack/docs/en/rest-api/security/get-builtin-privileges.asciidoc]: The doc test is currently disabled due to other feature flagged privileges

**Pull request merged**
